src/platform/GenericOTARequestorDriver.cpp

```diff
@@ -93,30 +152,162 @@ void GenericOTARequestorDriver::UpdateSuspended(System::Clock::Seconds32 delay)
+                                                      void * aAppState)
+{
+    SystemLayer().StartTimer(std::chrono::duration_cast<System::Clock::Timeout>(delay), action, aAppState);
+}
+
+void GenericOTARequestorDriver::CancelDelayedAction(System::TimerCompleteCallback action, void * aAppState)
+{
+    SystemLayer().CancelTimer(action, aAppState);
 }
 
-void GenericOTARequestorDriver::ScheduleDelayedAction(UpdateFailureState state, System::Clock::Seconds32 delay,
-                                                      System::TimerCompleteCallback action)
+void GenericOTARequestorDriver::ProcessAnnounceOTAProviders(
+    const ProviderLocationType & providerLocation,
+    app::Clusters::OtaSoftwareUpdateRequestor::OTAAnnouncementReason announcementReason)
 {
-    CHIP_ERROR error = SystemLayer().StartTimer(std::chrono::duration_cast<System::Clock::Timeout>(delay), action, this);
+    // If reason is URGENT_UPDATE_AVAILABLE, we start OTA immediately. Otherwise, respect the timer value set in mOtaStartDelayMs.
+    // This is done to exemplify what a real-world OTA Requestor might do while also being configurable enough to use as a test app.
+    uint32_t msToStart = 0;
+    switch (announcementReason)
+    {
+    case OTAAnnouncementReason::kSimpleAnnouncement:
+    case OTAAnnouncementReason::kUpdateAvailable:
+        msToStart = mOtaStartDelayMs;
+        break;
+    case OTAAnnouncementReason::kUrgentUpdateAvailable:
+        msToStart = kImmediateStartDelayMs;
+        break;
+    default:
+        ChipLogError(SoftwareUpdate, "Unexpected announcementReason: %u", static_cast<uint8_t>(announcementReason));
+        return;
+    }
+
+    // IMPLEMENTATION CHOICE:
+    // This implementation of the OTARequestor driver ignores the announcement if an update is in progress,
+    // otherwise it queries the provider passed in the announcement
 
-    if (error != CHIP_NO_ERROR)
+    if (mRequestor->GetCurrentUpdateState() != OTAUpdateStateEnum::kIdle)
     {
-        HandleError(state, error);
+        ChipLogProgress(SoftwareUpdate, "State is not kIdle, ignoring the AnnounceOTAProviders. State: %d",
+                        (int) mRequestor->GetCurrentUpdateState());
+        return;
     }
+
+    // Point the OTARequestor to the announced provider
+    mRequestor->SetCurrentProviderLocation(providerLocation);
+    mLastUsedProvider = providerLocation;
+
+    ScheduleDelayedAction(System::Clock::Seconds32(msToStart / 1000), StartDelayTimerHandler, this);
+}
+
+void GenericOTARequestorDriver::DriverSendQuery()
+{
+
+    // IMPLEMENTATION CHOICE
+    // In this implementation explicitly triggering a query cancels any in-progress update.
+    UpdateCancelled();
+
+    // Default provider timer only runs when there is no ongoing query/update; must stop it now.
+    // ConnectToProvider() will change the state from kIdle
+    StopDefaultProviderTimer();
+
+    mRequestor->ConnectToProvider(OTARequestorInterface::kQueryImage);
```

Comment:
I am still not clear on all the different ways to trigger a query from the driver. There is now:
- OTARequestor::ConnectToProvider
- OTARequestor::TriggerImmediateQuery
- OTARequestor::SendQuery

Looking at existing code, it doesn't even seems like `ConnectToProvider` should be exposed. This actually establishes the CASE session. We should reserve that to be used within `OTARequestor` class only. The driver should just make an API call that says "start a query", which would be one of the other two APIs above.

`TriggerImmediateQuery` just calls `DetermineAndSetProviderLocation` and then calls `SendQuery`. I don't see how that is ever needed when it is the caller/driver responsibility when it wants to query to have set this value already before issuing the request to query.

`SendQuery` currently still makes a call to `DriverSendQuery`

Reply:
Trying to see if we could simplify all the APis in both classes and just do:

Driver periodic query timer expires -> GenericOTARequestorDriver::DefaultProviderTimerHandler() which cancels current update and all timers, determine and set provider location -> OTARequestor::SendQuery() which changes the current state from Idle to Querying -> OTARequestor::ConnectToProvider()

For an AnnounceOTAProvider case:

OTARequestor::HandleAnnounceOTAProviders() -> GenericOTARequestorDriver::ProcessAnnounceOTAProviders() which sets the provider location -> schedule the StartTimer() -> StartDelayTimerHandler() -> OTARequestor::SendQuery() which changes the current state from Idle to Querying -> OTARequestor::ConnectToProvider()

We could get rid of an extra function in OTARequestor class (maybe `TriggerImmediateQuery`) and one in this class `DriverSendQuery`

Reply:
The only reason `TriggerImmediateQuery` and `SendQuery` exist are to be the application APIs -- a way for the application to say to the OTARequestor "Start an OTA query using whatever provider your choose" and "Start and OTA query to the provider that I have just set". 
I thought that we need APIs like this, but this is certainly up for discussion. In the very least the first one is needed, don't you think?

Reply:
I think the current implementation is reasonably streamlined for the three cases
DefaultProviderTimerHandler ->DriverSendQuery ->ConnectToProvider 
ProcessAnnounceOTAProviders->StartDelayTimerHandler->DriverSendQuery->ConnectToProvider
(and if we get rid of SendQuery)
TriggerImmediateQuery->DriverSendQuery->ConnectToProvider

Reply:
> The only reason TriggerImmediateQuery and SendQuery exist is to be the application APIs -- a way for the application to say to the OTARequestor "Start an OTA query using whatever provider your choose" and "Start an OTA query to the provider that I have just set".

I thought we had talked about the OTARequestor not making decisions like choosing a provider. That decision should be up to the caller, which is the second case you mention here.

> I thought that we need APIs like this, but this is certainly up for discussion. In the very least the first one is needed, don't you think?

Agreed that we need an API for the "Start an OTA query to the provider that I have just set", where "I' is the driver. The other case, goes against the whole idea that the driver is dictating what to do.

Reply:
I don't think `ConnectToProvider` should be called outside of OTARequestor class. This function establishes the CASE session. We really just need APIs that allows entities (like the driver) to say, hey, I want to start a query... so APIs like `SendQuery` or `TriggerImmediateQuery` is more appropriate.

Reply:
I also don't think DriverSendQuery is necessary. Specifically for this:
> TriggerImmediateQuery->DriverSendQuery->ConnectToProvider

This expanded to include the classes the functions belong to is OTARequestor::TriggerImmediateQuery->GenericOTARequestorDriver::DriverSendQuery-> OTARequestor ::ConnectToProvider

Why the need to go back and forth? Once the OTARequestor is told to trigger the query (probably by the dirver), it should just go and call `ConnectToProvider`. Why does it need to go to the driver in that case and then to end up coming back to the OTARequestor?

Reply:
> Why the need to go back and forth? Once the OTARequestor is told to trigger the query (probably by the dirver), it should just go and call `ConnectToProvider`. Why does it need to go to the driver in that case and then to end up coming back to the OTARequestor?

Because the it is the driver (i.e. implementation) that fist needs to execute the logic choices it's responsible for. That's what this PR is doing -- all the choices as to when to start/stop the periodic timer, whether to cancel the ongoing update are moved to the driver. In this particular implementation the periodic timer needs to be stopped and possible running update timers must be cancelled -- according to the above approach this can't be handled in the Requestor

Reply:
> I don't think `ConnectToProvider` should be called outside of OTARequestor class. This function establishes the CASE session. We really just need APIs that allows entities (like the driver) to say, hey, I want to start a query... so APIs like `SendQuery` or `TriggerImmediateQuery` is more appropriate.

Sure, we can introduce an API like that but it would just be another intermediate step that would do nothing but call `ConnectToProvider` . The `ConnectToProvider` API was introduced by @Damian-Nordic to replace a bunch of functions doing exactly that.

Reply:
> I thought we had talked about the OTARequestor not making decisions like choosing a provider. That decision should be up to the caller, which is the second case you mention here.

In TriggerImmediateQuery it is the driver that's making the decision in this call mOtaRequestorDriver->DetermineAndSetProviderLocation();

> Agreed that we need an API for the "Start an OTA query to the provider that I have just set", where "I' is the driver. The other case, goes against the whole idea that the driver is dictating what to do.

But this is an application API, here "I" is the application, not the driver.

Reply:
> Because the it is the driver (i.e. implementation) that first needs to execute the logic choices it's responsible for.

The driver already made the logic choice when it called `TriggerImmediateQuery`. Prior to calling this function, the driver should have cancelled it's timers and then ask for a query.

If I'm understanding correctly here, you're suggesting the driver call `TriggerImmediateQuery` and then have the requestor then tell the driver to go cancel it's timers, then the driver tells the requestor to query... that is completely redundant.

Reply:
> Sure, we can introduce an API like that but it would just be another intermediate step that would do nothing but call ConnectToProvider . The ConnectToProvider API was introduced by @Damian-Nordic to replace a bunch of functions doing exactly that.

Not introduce a new API. That is what `TriggerImmediateQuery` does underneath (at least existing code prior to your PR). It is the one that calls `ConnectToProvider(kQueryImage)` to indicate we are establishing a CASE session with the purpose to send a query image command once established.

Prior to your PR, the one file/class calling `ConnectToProvider` is the OTARequestor class.

Reply:
> But this is an application API, here "I" is the application, not the driver.

Can you please give me a scenario for which an application needs to send a query image directly? Also, regardless of the scenario, the driver should also call `DetermineAndSetProviderLocation` or have it's own impl of that prior to query. It should not ask the OTARequestor class to go determine a provider location to use.

Reply:
OK, I think I understand where our main disconnect is. 
With the approach taken in this PR it is not possible to have a single "initiate query" API that would be callable both from the application and from the driver. 
That API has to reside in OTARequestor. When called from the application this API has to invoke (at some point in the chain) the logic that cancels the ongoing update, stops the periodic timer and sets the Provider. When called from the driver it cannot do any of those things because they must happen before that API is called (because it can be called from several places where the choices are different and the knowledge of what has to be done will be lost after the API is entered (those places are: AnnounceOTAProviders, various retry scenarios, periodic timer firing)).

Reply:
>Can you please give me a scenario for which an application needs to send a query image directly? 

Just about any internet-connected consumer device I have seen has a "Check for updates" menu option.

Reply:
I agree @selissia that letting the application-level logic on a device be able to be triggered by user action to query was definitely an intended allowed method of query, independent of AnnounceOTAProvider and the "every 24h" logic